### PR TITLE
Port fontconfig's font weight detection to font_manager.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

This (should) fix the various font weight detection problems (#4822 #8551 #8607 #8788 #8800) by reusing fontconfig's font weight detection algorithm (specifically https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/452be81/src/fcfreetype.c#L1859 https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/452be81/src/fcfreetype.c#L1928 https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/452be81/src/fcfreetype.c#L2006 https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/452be81/src/fcfreetype.c#L2044).

The first commit is a straight port of fontconfig's algorithm, and outputs *fontconfig weights*; the second commit reverts to using CSS/OpenType weights, which is what we were using so far (see discussion in https://github.com/matplotlib/matplotlib/issues/10249).

To test this, delete your last fontlist-v310.json and regen one by importing matplotlib.font_manager with or without this PR *and* setting the PYTHONHASHSEED envvar to a fixed value -- this ensures that the json dumps stay in the same order, as we use sets, whose order depend on the hash seed.

(If we agree on the PR I'll need to bump the fontlist version number as well.)

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/font_manager.py [hunk 1]
```diff
@@ -384,10 +421,50 @@ def ttfFontProperty(font):
...
+                              sfnt.get((*mac_key, font_subfamily), b'')
+                              .decode('latin-1'),
+                              sfnt.get((*ms_key, wws_subfamily), b'')
+                              .decode('utf-16-be'),
+                              sfnt.get((*ms_key, typographic_subfamily), b'')
+                              .decode('utf-16-be'),
+                              sfnt.get((*ms_key, font_subfamily), b'')
+                              .decode('utf-16-be'),
+                              ))]
+              or [font.style_name])
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/font_manager.py]: yes

File: lib/matplotlib/font_manager.py [hunk 2]
```diff
@@ -384,10 +421,45 @@ def ttfFontProperty(font):
...
+        if os2 and os2["version"] != 0xffff:
+            return os2["usWeightClass"]
+        # PostScript font info weight.
+        try:
+            ps_font_info = font.get_ps_font_info()
+        except ValueError:
+            pass
+        else:
+            for regex, weight in _weight_regexes:
+                if re.fullmatch(regex, ps_font_info["weight"] or "", re.I):
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py:451]: This is the `FcIsWeight` in fontconfig, which calls `FcStringIsConst`, which calls `FcStrCmpIgnoreBlanksAndCase`, which ignores _any_ spaces. Do we actually normalize `ps_font_info`, or does that replacement need to be done here too?

File: lib/matplotlib/font_manager.py [hunk 3]
```diff
@@ -384,10 +421,45 @@ def ttfFontProperty(font):
...
+                    return weight
+        # Style name weight.
+        for style in styles:
+            for regex, weight in _weight_regexes:
+                if re.search(regex, style, re.I):
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py:456]: Same thing here, except it's `FcContainsWeight`, and `FcStringContainsConst`.

**Comment by QuLogic** [comment]
I tried with Fira Sans, which has a bunch of styles, and it went from (heaviest to lightest):
```
FiraSans-Black black 900
FiraSans-BlackItalic black 900
FiraSans-Bold 700
FiraSans-BoldItalic 700
FiraSans-ExtraBold bold 700
FiraSans-ExtraBoldItalic bold 700
FiraSans-SemiBold semibold 600
FiraSans-SemiBoldItalic semibold 600
FiraSans-Medium medium 500
FiraSans-MediumItalic medium 500
FiraSans-Regular regular 400
FiraSans-Italic 400
FiraSans-Thin 400
FiraSans-ThinItalic 400
FiraSans-Light light 200
FiraSans-LightItalic light 200
FiraSans-ExtraLight light 200
FiraSans-ExtraLightItalic light 200
```
to:
```
FiraSans-Black 900
FiraSans-BlackItalic 900
FiraSans-ExtraBold 800
FiraSans-ExtraBoldItalic 800
FiraSans-Bold 700
FiraSans-BoldItalic 700
FiraSans-SemiBold 600
FiraSans-SemiBoldItalic 600
FiraSans-Medium 500
FiraSans-MediumItalic 500
FiraSans-Regular 400
FiraSans-Italic 400
FiraSans-Light 300
FiraSans-LightItalic 300
FiraSans-ExtraLight 275
FiraSans-ExtraLightItalic 275
FiraSans-Thin 250
FiraSans-ThinItalic 250
```
This is certainly better in terms of distinct values, and it matches the order shown on Google Fonts.  However, Google Fonts calls ExtraLight 200, and Thin 100.

File: lib/matplotlib/font_manager.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/font_manager.py:451]: I probably just missed that, fixed accordingly.

File: lib/matplotlib/font_manager.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/font_manager.py:456]: ditto.

**Comment by anntzer** [comment]
This is what I get:
```
python -c 'from matplotlib.font_manager import ft2font, ttfFontProperty; from pathlib import Path; [print(path, ttfFontProperty(ft2font.FT2Font(str(path))).weight) for path in Path("/usr/share/fonts/TTF").glob("FiraSans-*.ttf")]' | sort -n -k2
/usr/share/fonts/TTF/FiraSans-EightItalic.ttf 100
/usr/share/fonts/TTF/FiraSans-Eight.ttf 100
/usr/share/fonts/TTF/FiraSans-FourItalic.ttf 100
/usr/share/fonts/TTF/FiraSans-Four.ttf 100
/usr/share/fonts/TTF/FiraSans-HairItalic.ttf 100
/usr/share/fonts/TTF/FiraSans-Hair.ttf 100
/usr/share/fonts/TTF/FiraSans-ThinItalic.ttf 100
/usr/share/fonts/TTF/FiraSans-Thin.ttf 100
/usr/share/fonts/TTF/FiraSans-TwoItalic.ttf 100
/usr/share/fonts/TTF/FiraSans-Two.ttf 100
/usr/share/fonts/TTF/FiraSans-ExtraLightItalic.ttf 200
/usr/share/fonts/TTF/FiraSans-ExtraLight.ttf 200
/usr/share/fonts/TTF/FiraSans-UltraLightItalic.ttf 200
/usr/share/fonts/TTF/FiraSans-UltraLight.ttf 200
/usr/share/fonts/TTF/FiraSans-LightItalic.ttf 300
/usr/share/fonts/TTF/FiraSans-Light.ttf 300
/usr/share/fonts/TTF/FiraSans-BookItalic.ttf 350
/usr/share/fonts/TTF/FiraSans-Book.ttf 350
/usr/share/fonts/TTF/FiraSans-Italic.ttf 400
/usr/share/fonts/TTF/FiraSans-Regular.ttf 400
/usr/share/fonts/TTF/FiraSans-MediumItalic.ttf 500
/usr/share/fonts/TTF/FiraSans-Medium.ttf 500
/usr/share/fonts/TTF/FiraSans-SemiBoldItalic.ttf 600
/usr/share/fonts/TTF/FiraSans-SemiBold.ttf 600
/usr/share/fonts/TTF/FiraSans-BoldItalic.ttf 700
/usr/share/fonts/TTF/FiraSans-Bold.ttf 700
/usr/share/fonts/TTF/FiraSans-ExtraBoldItalic.ttf 800
/usr/share/fonts/TTF/FiraSans-ExtraBold.ttf 800
/usr/share/fonts/TTF/FiraSans-HeavyItalic.ttf 900
/usr/share/fonts/TTF/FiraSans-Heavy.ttf 900
/usr/share/fonts/TTF/FiraSans-UltraItalic.ttf 950
/usr/share/fonts/TTF/FiraSans-Ultra.ttf 950
```
that looks correct?

**Comment by QuLogic** [comment]
I pulled these fonts from [Google Fonts](https://fonts.google.com/specimen/Fira+Sans), but maybe they're wrong?

The packaged version are .otf, so I'm not sure if they're equivalent, but if I use those, the weights are  the same as yours, except for ExtraLight, which is 250 instead of 200.

If I run `fc-scan FiraSans-Thin.{ttf,otf}`, then the otf has `weight:0` while the ttf has `weight:45.5`.

**Comment by anntzer** [comment]
Mine were from Arch's distro packages.
I agree with your observations re google's fonts.  However the weight values are just directly read (in this case) from the OS/2 table's usWeightClass entry (https://docs.microsoft.com/en-us/typography/opentype/spec/os2#usweightclass), which is the first thing FreeType looks at.  So I'm quite tempted to just claim that google's fonts have incorrect metadata...

**Comment by QuLogic** [comment]
I also confirmed in FontForge, so I guess I'll report a but on Google Fonts.

File: lib/matplotlib/font_manager.py [hunk 4]
```diff
@@ -384,10 +421,46 @@ def ttfFontProperty(font):
...
+        except ValueError:
+            pass
+        else:
+            for regex, weight in _weight_regexes:
+                if re.fullmatch(regex, ps_font_info_weight, re.I):
+                    return weight
+        # Style name weight.
+        for style in styles:
+            for regex, weight in _weight_regexes:
+                if re.search(regex, style.replace(" ", ""), re.I):
```
Comment (QuLogic) [inline lib/matplotlib/font_manager.py:457]: Could also hoist the `replace` into the outer loop, if you want.

Comment (anntzer) [reply lib/matplotlib/font_manager.py]: sure

**Comment by QuLogic** [comment]
> (If we agree on the PR I'll need to bump the fontlist version number as well.)

Don't forget this.

**Comment by anntzer** [comment]
done

**Comment by jklymak** [comment]
@tacaswell you self-requested a review on this.  Did you want it to wait?  

I'd merge on the basis of @anntzer and @QuLogic combined expertise, and because it seems an improvement, not because I particularly understand the guts of the font weight detection.

**Comment by tacaswell** [comment]
If you think it is ok, merge it.

**Comment by AlvaroFurlani** [comment]
> 

Is there any resolution for Google Fonts? I'm trying to use Public Sans and the weight in matplotlib for the Thin and ExtraLight styles are both 250, instead of 100 and 200, respectively.

**Comment by QuLogic** [comment]
Please open a new issue with an explanation of what you mean.

**Pull request merged**